### PR TITLE
Move getting container name to _profile_process function, expand ProfileData with container_name property

### PR DESCRIPTION
**Pull request by mpozniak95**
## Description
Get container name for profiled process in the same function as we get appid and metadata (_profile_process).

## Related Issue
https://github.com/Granulate/gprofiler/issues/640

### CODE REVIEW

File: gprofiler/profilers/php.py [hunk 1]
```diff
@@ -66,9 +67,12 @@ def __init__(
         profile_spawned_processes: bool,
         php_process_filter: str,
         php_mode: str,
+        container_names_client: Optional[ContainerNamesClient],
     ):
         assert php_mode == "phpspy", "PHP profiler should not be initialized, wrong php_mode value given"
-        super().__init__(frequency, duration, stop_event, storage_dir, insert_dso_name, profiling_mode)
+        super().__init__(
+            frequency, duration, stop_event, storage_dir, insert_dso_name, profiling_mode, container_names_client
+        )
```
Comment (Jongy) [inline gprofiler/profilers/php.py:75]: I don't like it much that we have to make this change in all profilers.

Logically - there are profiler parameters which are truly parameters per profiler and might change per profiler - `frequency`, `duration`. All others (and specifically `stop_event`, `storage_dir`, `containers_name_client`) are actually passed because we cannot pass "general state parameters" in another way.

Let's refactor it (in a different PR) and then base this PR on that refactor.

* One option - profilers can get the `GProfiler` object and access the settings configured on it via accessors `.get_storage_dir()` etc.
* We can create another object `ProfilerState`, create it in `GProfiler` and pass it in the c'tor

My point is, adding a new "state parameter" like `containers_name_client` should not induce diff in all classes of all c'tors.

Let's open another PR for this refactor please

File: gprofiler/profilers/python.py [hunk 2]
```diff
@@ -215,6 +224,10 @@ def _profile_process(self, process: Process, duration: int, spawned: bool) -> Pr
         )
         appid = application_identifiers.get_python_app_id(process)
         app_metadata = self._metadata.get_metadata(process)
+        if self._container_names_client:
+            container_name = self._container_names_client.get_container_name(process.pid)
+        else:
+            container_name = ""
```
Comment (Jongy) [inline gprofiler/profilers/python.py:230]: This logic repeats itself - let's encapsulate to a helper. This helper can be a part of the refactor I suggested in https://github.com/Granulate/gprofiler/pull/667/files#r1091286777 - the `ProfilerState`, or `GProfiler`, can expose a method `get_container_name(pid: int)` which gets you the name or `""`.

File: gprofiler/merge.py [hunk 3]
```diff
@@ -311,13 +315,10 @@ def _enrich_pid_stacks(
         application_prefix = f"{idx};"
     else:
         application_prefix = ""
-
```
Comment (Jongy) [inline gprofiler/merge.py:314]: Bring it back :) Refrain from unrelated formatting changes

**Review by Jongy — COMMENTED** [review]
Overall looks good. I gave some comments about the design

**Comment by Jongy** [comment]
Please merge from master after https://github.com/Granulate/gprofiler/pull/668, then continue this PR

File: gprofiler/main.py [hunk 4]
```diff
@@ -267,6 +268,7 @@ def stop(self) -> None:
             prof.stop()
 
     def _snapshot(self) -> None:
+        self._container_names_client.refresh_container_names_cache()
```
[the remark this answers is not in the document]
Comment (mpozniak95) [reply gprofiler/main.py]: I think that we need to refresh container cache at the beginning of profiling. If profiled process will be terminated during profiling we won't be able to get his container name because there might be a situation when it will happen before first call of `get_container_name()` and there won't be any container_pid's in cache. I added a test to test_sanity.py to simulate such case.

File: gprofiler/containers_client.py [hunk 5]
```diff
@@ -49,6 +49,7 @@ def get_container_name(self, pid: int) -> str:
             return ""
 
         if pid in self._pid_to_container_name_cache:
+            self._current_container_names.add(self._pid_to_container_name_cache[pid])
             return self._pid_to_container_name_cache[pid]
```
Comment (Jongy) [inline gprofiler/containers_client.py:53]:
```suggestion
            name = self._pid_to_container_name_cache[pid]
            self._current_container_names.add(name)
            return name
```

nicer w/o accessing it twice

File: gprofiler/profiler_state.py [hunk 6]
```diff
@@ -40,3 +44,17 @@ def insert_dso_name(self) -> bool:
...
+
+    def refresh_container_cache(self) -> None:
+        if self._container_names_client:
+            self._container_names_client.refresh_container_names_cache()
+
+    def get_container_name(self, pid: int) -> str:
+        if self._container_names_client:
+            return self._container_names_client.get_container_name(pid)
+        else:
+            return ""
```
Comment (Jongy) [inline gprofiler/profiler_state.py:60]: Any particular reason you moved these wrappers into this class? We only use it to hold the reference - no need to clone methods into it.

if you did it as a way for code sharing (a wrapper, non-None class for the containers client) then it's okay

File: gprofiler/profiler_state.py [hunk 7]
```diff
@@ -40,3 +44,17 @@ def insert_dso_name(self) -> bool:
     @property
     def profiling_mode(self) -> str:
         return self._profiling_mode
+
+    @property
+    def container_names_client(self) -> Optional[ContainerNamesClient]:
+        return self._container_names_client
+
+    def refresh_container_cache(self) -> None:
+        if self._container_names_client:
```
Comment (Jongy) [inline gprofiler/profiler_state.py:53]:
```suggestion
        if self._container_names_client is not None:
```

Please fix everywhere - no implicit `None`/bool conversions

File: gprofiler/profilers/perf.py [hunk 8]
```diff
@@ -354,7 +354,8 @@ def _generate_profile_data(self, stacks: StackToSampleCount, pid: int) -> Profil
             appid = self._get_appid(pid)
         else:
             appid = None
-        return ProfileData(stacks, appid, metadata)
+        container_name = self._profiler_state.get_container_name(pid)
+        return ProfileData(stacks, appid, metadata, container_name)
```
Comment (Jongy) [inline gprofiler/profilers/perf.py:358]:
```suggestion
        return ProfileData(stacks, appid, metadata, self._profiler_state.get_container_name(pid))
```

No need for the extra variable, if it's only used once

File: gprofiler/profilers/java.py [hunk 9]
```diff
@@ -789,7 +789,7 @@ def _disable_profiling(self, cause: str) -> None:
             self._safemode_disable_reason = cause
 
     def _profiling_skipped_profile(self, reason: str, comm: str) -> ProfileData:
-        return ProfileData(self._profiling_error_stack("skipped", reason, comm), None, None)
+        return ProfileData(self._profiling_error_stack("skipped", reason, comm), None, None, None)
```
Comment (Jongy) [inline gprofiler/profilers/java.py:792]: This change (generally wherever you pass `None` as container name for a process that we don't profile due to skip / error, but the process is still alive) - because past code would have added the container name in `merge.py`.

Now, you add the name in `perf.py` and in e.g `java.py` for a java process. Then during merging, we select the java `ProfileData`. If the java was `Profiling skipped`, it will have no container name, but the perf one will have!

I suggest: if the process `ProfileData` has no name, but the perf one has, take the perf. This keeps the past behavior for processes we don't profile with a runtime profiler, and gives your enhanced behavior for processes that we do profile with e.g `JavaProfiler`.

File: tests/test_sanity.py [hunk 10]
```diff
@@ -241,3 +243,33 @@ def test_from_container_spawned_process(
         assert_collapsed(collapsed)
 
     assert b"Profiling spawned process" in container.logs()
+
+
+@pytest.mark.parametrize("in_container", [True])
+@pytest.mark.parametrize("profiler_type", ["attach-maps"])
+@pytest.mark.parametrize("runtime", ["nodejs"])
+@pytest.mark.parametrize("application_image_tag", ["without-flags"])
+def test_container_name_when_stopped(
```
Comment (Jongy) [inline tests/test_sanity.py:252]: Nice test! Good job coming up with it :)

To improve - let's make it non-racey, currently it's racey and might "pass" despite the app not actually going down before the profiler did. Try to write tests that are self-guarding, i.e if you have any assumptions on the flow, then you can guard them.

One way I suggest to do it is to profile a Java app instead of nodejs. For Java, we have logs that indicate a process went down during profiling - `logger.debug(f"Profiled process {ap_proc.process.pid} exited before stopping async-profiler")`.
Your test can run the Java, begin profiling, kill the Java and ensure that this log message appears.
Also, as a best practice let's use `assert_collapsed` here to make sure the profiling of the app really worked.

File: tests/test_sanity.py [hunk 11]
```diff
@@ -241,3 +243,33 @@ def test_from_container_spawned_process(
...
+    profiler_type: str,
+    profiler_flags: List[str],
+    application_docker_container: Container,
+) -> None:
+    profiler_flags.extend(["-d", "20"])
+    container = start_gprofiler_in_container_for_one_session(
+        docker_client, gprofiler_docker_image, output_directory, output_collapsed, runtime_specific_args, profiler_flags
+    )
+    try:
+        wait_event(20, Event(), lambda: re.search(rb"Started perf", container.logs()) is not None)
```
Comment (Jongy) [inline tests/test_sanity.py:267]: After you cange to Java you can fix this log searching for the log message of profiling a java process.

File: tests/test_sanity.py (hunk 10 above)
Comment (Jongy) [inline tests/test_sanity.py:252]: I'd add some docstring explaining the test and linking to the GH ticket. Something short like here:
```
@pytest.mark.parametrize("in_container", [True])
def test_java_jattach_async_profiler_log_output(
    application_pid: int,
    caplog: LogCaptureFixture,
    profiler_state: ProfilerState,
) -> None:
    """
    Tests that AP log is collected and logged in gProfiler's log.
    """
    ...
```

File: gprofiler/merge.py [hunk 12]
```diff
@@ -297,6 +285,11 @@ def _enrich_pid_stacks(
     """
     # generate application name
     appid = profile.appid
+    if profile.container_name:
+        container_prefix = profile.container_name
+    else:
+        container_prefix = ""
+    container_prefix = f"{container_prefix};"
```
Comment (Jongy) [inline gprofiler/merge.py:292]: Can extract this to a function

File: gprofiler/main.py [hunk 13]
```diff
@@ -142,18 +142,19 @@ def __init__(
         # 2. accessible only by us.
         # the latter can be root only. the former can not. we should do this separation so we don't expose
         # files unnecessarily.
+        if self._enrichment_options.container_names:
+            container_names_client: Optional[ContainerNamesClient] = ContainerNamesClient()
+        else:
+            container_names_client = None
```
Comment (Jongy) [inline gprofiler/main.py:148]: Can be a short `ContainerNamesClient() if ... else None` in the call to the ProfilerState c'tor

File: gprofiler/profiler_state.py (hunk 6 above)
Comment (Jongy) [reply gprofiler/profiler_state.py:60]: Haha I see that it was my own suggestion :sweat_smile: 

>  the ProfilerState, or GProfiler, can expose a method get_container_name(pid: int) which gets you the name or "".

I'm trying to avoid from making ProfilerState a god object. But alright let's go with it here, it's useful.

File: gprofiler/containers_client.py [hunk 14]
```diff
@@ -89,8 +90,11 @@ def _get_container_name(self, container_id: str) -> Optional[str]:
...
         return container_name
 
-    def _refresh_container_names_cache(self) -> None:
+    def refresh_container_names_cache(self) -> None:
         # We re-fetch all of the currently running containers, so in order to keep the cache small we clear it
         self._container_id_to_name_cache.clear()
+        self._pid_to_container_name_cache.clear()
         for container in self._containers_client.list_containers() if self._containers_client is not None else []:
             self._container_id_to_name_cache[container.id] = container.name
+            if container.pid:
```
Comment (Jongy) [inline gprofiler/containers_client.py:99]: is not None?

File: gprofiler/main.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Jongy) [reply gprofiler/main.py]: I see. Something is still off here for me, I will re-review thoroughly after you fix the other points I commented now

**Review by Jongy — CHANGES_REQUESTED** [review]
Done reviewing, looks very good :)

File: gprofiler/main.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mpozniak95) [reply gprofiler/main.py]: This is not needed, I deleted this and check for every process at the beginning of perf snapshot

File: gprofiler/profilers/dotnet.py [hunk 15]
```diff
@@ -93,6 +93,7 @@ def _make_command(self, process: Process, duration: int, output_path: str) -> Li
         ]
 
     def _profile_process(self, process: Process, duration: int, spawned: bool) -> ProfileData:
+        container_name = self._profiler_state.get_container_name(process.pid)
```
Comment (Jongy) [inline gprofiler/profilers/dotnet.py:96]: Nice next to the app_metadata

Comment (Jongy) [reply gprofiler/profilers/dotnet.py:96]: (in all profilers code, to keep it organized)

File: gprofiler/profilers/java.py [hunk 16]
```diff
@@ -902,6 +902,7 @@ def _check_async_profiler_loaded(self, process: Process) -> bool:
         return False
 
     def _profile_process(self, process: Process, duration: int, spawned: bool) -> ProfileData:
+        container_name = self._profiler_state.get_container_name(process.pid)
```
Comment (Jongy) [inline gprofiler/profilers/java.py:905]: Can move next to app_metadata and appid as suggested https://github.com/Granulate/gprofiler/pull/667/files#r1116345068

File: gprofiler/profilers/java.py [hunk 17]
```diff
@@ -957,8 +958,7 @@ def _profile_process(self, process: Process, duration: int, spawned: bool) -> Pr
             self._report_meminfo,
         ) as ap_proc:
             stackcollapse = self._profile_ap_process(ap_proc, comm, duration)
-
```
Comment (Jongy) [inline gprofiler/profilers/java.py:960]: Undo this change (no unrelated formatting changes)

File: gprofiler/profilers/python.py [hunk 18]
```diff
@@ -202,6 +202,7 @@ def _make_command(self, pid: int, output_path: str, duration: int) -> List[str]:
         return command
 
     def _profile_process(self, process: Process, duration: int, spawned: bool) -> ProfileData:
+        container_name = self._profiler_state.get_container_name(process.pid)
```
Comment (Jongy) [inline gprofiler/profilers/python.py:205]: Next to appid & app_metadata https://github.com/Granulate/gprofiler/pull/667/files#r1116345068

File: gprofiler/profilers/ruby.py [hunk 19]
```diff
@@ -99,6 +99,7 @@ def _make_command(self, pid: int, output_path: str, duration: int) -> List[str]:
         ]
 
     def _profile_process(self, process: Process, duration: int, spawned: bool) -> ProfileData:
+        container_name = self._profiler_state.get_container_name(process.pid)
```
Comment (Jongy) [inline gprofiler/profilers/ruby.py:102]: Next to appid and app metadata https://github.com/Granulate/gprofiler/pull/667/files#r1116345068

File: tests/test_sanity.py [hunk 20]
```diff
@@ -241,3 +243,42 @@ def test_from_container_spawned_process(
...
+    """
+    profiler_flags.extend(["-d", "20"])
+    container = start_gprofiler_in_container_for_one_session(
+        docker_client, gprofiler_docker_image, output_directory, output_collapsed, runtime_specific_args, profiler_flags
+    )
+    try:
+        wait_event(
+            20,
+            Event(),
+            lambda: re.search(rb"Profiling .*process \d* with async-profiler", container.logs()) is not None,
```
Comment (Jongy) [inline tests/test_sanity.py:274]: I know it was my suggestion but I realized it makes no sense to use a Java app here - once the container is dead, async-profiler dies too, as it lives in the Java app... so how can we read its output (it writes the output in the container, but it's dead)? lol. Did you get this test to pass? Because I think it should not.

Another suggestion - let's use a Python app with `profiler_type=py-spy` - it prints an indicative message upon process exit, that is dumped into the profiler log:
```
py-spy> Stopped sampling because process exited
```

You can kill the app, then search for this message.

As a side note, to ensure this test covers exactly the case we want to test, you can build gProfiler from master and run it - it should fail on master and pass on your branch.

**Review by Jongy — CHANGES_REQUESTED** [review]
Done reviewing, one comment about the test and a few other small

File: tests/test_sanity.py (hunk 20 above)
Comment (mpozniak95) [reply tests/test_sanity.py:274]: Sure I will change it to py-spy. I was able to pass both java and nodeJs test on this PR and I tested both on master and both failed :)

Comment (mpozniak95) [reply tests/test_sanity.py]: Changed, Also works on this PR fails on master

File: gprofiler/profilers/java.py [hunk 21]
```diff
@@ -938,8 +941,6 @@ def _profile_process(self, process: Process, duration: int, spawned: bool) -> Pr
             self._profiled_pids.add(process.pid)
 
         logger.info(f"Profiling{' spawned' if spawned else ''} process {process.pid} with async-profiler")
-        app_metadata = self._metadata.get_metadata(process)
```
Comment (Jongy) [inline gprofiler/profilers/java.py:941]: This was the correct place for app metadata & appid retrieval, I don't want them to be retrieved before the safety checks are done (async-profiler loaded, JVM profiling supported etc). Bring them back and move the container name here as well?

File: tests/test_sanity.py (hunk 20 above)
Comment (Jongy) [reply tests/test_sanity.py]: Awesome

**Review by Jongy — CHANGES_REQUESTED** [review]
One final comment.

I went over existing tests and noticed we don't currently have any tests about container names. You added one here for runtime profilers, but we changed the way container names are added to perf-profiled processes, and that's not tested yet. Can you add testing of container names to a perf test, e.g to `test_perf_fp_dwarf_smart`? You can get the name the same way with `application_docker_container` and ensure it's present in the stack trace (you can also create a helper function `assert_container_names` to be used in both tests)

**Comment by mpozniak95** [comment]
Since we are enriching stack during merging profiles in `main.py` and in `test_perf_fp_dwarf_smart` we use only SystemProfiler object, I added checking if container_name is in process ProfileData instead of checking if it is in stack trace. Is this enough?

File: tests/test_perf.py [hunk 22]
```diff
@@ -67,6 +68,8 @@ def test_perf_fp_dwarf_smart(
         assert is_function_in_collapsed(";_start;__libc_start_main;main;", process_collapsed) ^ bool(
             perf_mode not in ("dwarf", "smart")
         )
+        # Check if container name is added to ProfileData
+        assert application_docker_container.name == snapshot_pid_profile(profiler, application_pid).container_name
```
Comment (Jongy) [inline tests/test_perf.py:72]: Avoid calling `snapshot` again - use `snapshot_pid_profiler` in line 54, then compare both the container name & the collapsed of it. Calling `snapshot` again invokes the profiler twice, unnecessarily.

**Comment by Jongy** [comment]
One comment :)

**Review by Jongy — APPROVED** [review]
